Update command

xh3b4sd opened the pull request:
See https://github.com/giantswarm/giantswarm/issues/548

<!---
@huboard:{"custom_state":"archived"}
-->

Commit:
added basic update command

Commit:
Merge branch 'master' into update-command

Commit:
added basic update command

Commit:
Merge branch 'master' into update-command

Commit:
make `make lint` happy

xh3b4sd commented:
This is the first draft out of my head, based on the discussions we had recently. Please provide some feedback. Comments, tests and all the other stuff will follow later in case the basic idea is right. 

RFC @zeisss @JosephSalisbury @Nesurion

zeisss commented on cli/update.go:
```diff
@@ -0,0 +1,62 @@
+	maybeBlockWithFeedback(blockWithFeedbackCtx{
+		Request:    req,
+		Descriptor: "update",
+		NoBlock:    globalFlags.NoBlock,
```
not sure we want to allow `noBlock` here, as it would currently break the update, since the CLI would quit.

xh3b4sd replied:
Good point.

zeisss commented on controller/controller.go:
```diff
@@ -65,6 +67,14 @@ func DefaultConfig() Config {
+	// found, Inago assumes the whole group slice to be "dirty" and returns true
+	// having the group slices removed from the given req that are up to date,
+	// otherwise false, leaving the req as it is.
+	GroupNeedsUpdate(req Request) (Request, bool, error)
```
:one: The description suggests I get the slices that do not need to be changed. The implementation looks like I only get the dirty ones. I also find the behavior to return the original array when nothing changed surprising. Can't we just return the dirty slices? If they are returned, they are all dirty. It's the client responsability to make a new request out of this.

:two: Why do we want this in the public interface?

xh3b4sd replied:
1) The comment states this.

> and returns true having the group slices removed from the given req that are up to date

The slices that are up to date will be removed. The slices that need to be updated stay in there. 

> Can't we just return the dirty slices?

Sure. That makes more sense. 

2) You are asking over and over again why this and that should be in the public interface. We already talked about this. Lets have a talk again so we don't need to talk about it in every pull request.

zeisss commented on controller/status.go:
```diff
@@ -90,6 +105,21 @@ func allStatesEqual(usl []fleet.UnitStatus) bool {
+			return false, maskAny(err)
+		}
+
+		if aggregated != StatusRunning {
```
Don't we want to use `status` here?

xh3b4sd replied:
Hehe, good catch. We want. Will do.

zeisss commented on controller/update.go:
```diff
@@ -0,0 +1,255 @@
+package controller
```
new file, yes! :cake:

zeisss commented on controller/update.go:
```diff
@@ -0,0 +1,255 @@
+// process. Lets have a look at how the update process of 3 group slices would
+// look like using the given options.
+//
+//     TODO I am not that happy with this visualization. Improving it? Removing it?
```
It gives a hint. Not perfect, but it helps. We can draw a png add it to the docs to make it better.

xh3b4sd replied:
Yeah getting the time stream right is not really possible here. At least I don't know how.

zeisss commented on controller/update.go:
```diff
@@ -0,0 +1,255 @@
+	// MaxGrowth represents the number of groups allowed to be added at a given
+	// time. No more than MaxGrowth groups will be added at the same time during
+	// the update process.
+	MaxGrowth int
```
which reminds me: We never actually discussed if we want this possibility of kubernetes to also specify percentage values for this. I guess if we want this, we can do it in a later PR. We should make an issue for this.

xh3b4sd replied:
See https://github.com/giantswarm/inago/issues/62

zeisss commented on controller/update.go:
```diff
@@ -0,0 +1,255 @@
+			}(jobDesc)
+		case err := <-fail:
+			close(done)
+			close(addQueue)
```
only writes should close a queue, not reads. this may lead to panics or worse.

xh3b4sd replied:
Good point. Which writer should close the channel in case there are multiple writers using it? Can't it happen that the panic happens there as well? So this would mean each writer needs to check if the channel is writeable. How to do so? 

Further the scope of this channels is `UpdateWithStrategy`. They are created and used only here. 

So what to do? Simply not closing them?

zeisss replied:
> Which writer should close the channel in case there are multiple writers using it?

I would say no one. A signal (cancel chan?) should tell those writes to stop and when all are stopped, some upper layer can close all channels. The upper layer would still be this function here, but we need to ensure that the workers are stopped

zeisss replied:
IIRC you can check if a write is possible with a `select/case`

xh3b4sd replied:
> we need to ensure that the workers are stopped

IMO this happens when we simply return here. That is when the done channel is called.

zeisss replied:
but we have to stop them first, then close the channel. not the other way around

xh3b4sd replied:
I would not close the channels then. Just end the for loop when the done channel was used by the "remove worker". Then there is nobody writing to the channels and we return.

zeisss commented on controller/update.go:
```diff
@@ -0,0 +1,255 @@
+	return false, nil
+}
+
+func (c controller) addWorker(jobDesc updateJobDesc, removeQueue chan updateJobDesc, fail chan<- error) {
```
can we name this `runAddWorker`? I was confused because I assumed we add a worker here, not do the "add" steps

xh3b4sd replied:
Sure. Will do.

zeisss commented on controller/update.go:
```diff
@@ -0,0 +1,255 @@
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
```
TODO: Wait for startup time to be reached?

xh3b4sd replied:
Good catch. Will add.

Commit:
heavy WIP

Commit:
WIP

Commit:
WIP

Commit:
Merge branch 'master' into update-command

Commit:
better job queuing

xh3b4sd commented:
RFR @JosephSalisbury @zeisss @Nesurion 

Today I had a tough time to get the update job queueing nice. Note that I changed the slice ID handling here already. So please have a look. I require at least one WFM (Works for me). So please check the code and play around with the submit/start/update commands.

xh3b4sd commented:
Note that the tests and comments are not cared about. Lets get the basics right first.

Nesurion commented:
Tried the following: create a single slice of a group.
ran an update with `inagoctl update -v --max-growth 0 --min-alive 2 --ready-secs 5 mygroup`. I know that it can't work with those parameters. But I figured I'd get an error message saying something like: `failed to update: invalid parameter combination.`  or something like that, maybe a bit more detail.
What happens is: inagoctl is stuck

Commit:
min alive validation

xh3b4sd commented:
Good catch. I added some validation. Please try again.

Nesurion commented:
Testing what happens when new units enter inactive, dead state in the ready-secs period.
What I did: 
`$ inagoctl submit mygroup 2`
`Succeeded to submit 2 slices for group 'mygroup': [5be cea].`

`$ inagoctl start mygroup`
`Succeeded to start 2 slices for group 'mygroup': [5be cea].``

`$ inagoctl update -v --max-growth 2 --min-alive 1 mygroup`

`$ fleetctl list-units`
old units are running. new units are started. entering ready-secs period.
new units:

```
mygroup-bar@010.service 087296d2.../172.17.8.101        active          running
mygroup-bar@2bd.service a8e6206f.../172.17.8.103        active          running
mygroup-foo@010.service a8e6206f.../172.17.8.103        active          running
mygroup-foo@2bd.service ebf055c8.../172.17.8.102        active          running
```

`$ fleetctl stop mygroup-bar@010.service`
stopping a units, to simulate a unit dying for whatever reason.

```
mygroup-bar@010.service 087296d2.../172.17.8.101        inactive        dead
mygroup-bar@2bd.service a8e6206f.../172.17.8.103        active          running
mygroup-foo@010.service a8e6206f.../172.17.8.103        active          running
mygroup-foo@2bd.service ebf055c8.../172.17.8.102        active          running
```

`Succeeded to update 2 slices for group 'mygroup': [010 2bd].`

=> inago does not stop when a newly started unit fails during the ready-secs period.

Commit:
check state during sleep-sec wait

xh3b4sd commented:
Nice one. Check the updated version. Works now with throwing an error in case the group is not running after the update.

Nesurion commented:
nice, both fixed! WFM

JosephSalisbury commented on common/common.go:
```diff
@@ -57,3 +44,17 @@ func sliceSuffix(name string) (string, error) {
+// everything up to the @ character, if any, or the given name without
+// extension.
+//
+//   app@1.service  =>  app
```
It might be an idea to use the "example functions" available in https://golang.org/pkg/testing/

xh3b4sd replied:
It is only one pull request away @JosephSalisbury!

JosephSalisbury replied:
https://github.com/giantswarm/inago/issues/77 <3

JosephSalisbury commented:
I did some fiddling:

```
core@core-01 ~ $ ls -R
.:
inagoctl  test-group

./test-group:
test-group-unit.service
core@core-01 ~ $ cat ./test-group/test-group-unit.service
[Unit]
Description=Inago Test Unit

[Service]
ExecStart=/bin/bash -c "while true; do echo Hi; sleep 10; done"
core@core-01 ~ $ fleetctl list-units
UNIT    MACHINE ACTIVE  SUB
core@core-01 ~ $ ./inagoctl submit test-group
Failed to submit 1 slices for group 'test-group': [d71]. (unit not found)
core@core-01 ~ $ fleetctl list-units
UNIT            MACHINE             ACTIVE      SUB
test-group-unit.service b944ede1.../172.17.8.101    inactive    dead
```

I would not expect the message on submitting the group - it seems like the group was submitted successfully.

```
core@core-01 ~ $ ./inagoctl status test-group
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0x84a600, 0xc820010080)
    /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/giantswarm/inago/common.SliceID(0xc8200b3340, 0x17, 0x0, 0x0, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/common/common.go:33 +0x137
github.com/giantswarm/inago/fleet.fleet.createOurStatusList(0xc82000b920, 0x8d67e0, 0x4, 0x0, 0x0, 0x0, 0x8e1770, 0xb, 0x0, 0x0, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/fleet/fleet.go:314 +0x119
github.com/giantswarm/inago/fleet.fleet.GetStatusWithMatcher(0xc82000b920, 0x8d67e0, 0x4, 0x0, 0x0, 0x0, 0x8e1770, 0xb, 0x0, 0x0, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/fleet/fleet.go:292 +0x6e9
github.com/giantswarm/inago/fleet.(*fleet).GetStatusWithMatcher(0xc82007a6c0, 0xc820012eb0, 0x0, 0x0, 0x0, 0x0, 0x0)
    <autogenerated>:7 +0xe3
github.com/giantswarm/inago/controller.controller.getExistingSliceIDs(0x7f4dae73c9a8, 0xc82007a6c0, 0x7f4dae73cab8, 0xb39680, 0x7f4dae73ca30, 0xc8200b3000, 0x3, 0x3b9aca00, 0x45d964b800, 0x7ffc1f506de1, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/controller/request.go:102 +0xee
github.com/giantswarm/inago/controller.controller.ExtendWithExistingSliceIDs(0x7f4dae73c9a8, 0xc82007a6c0, 0x7f4dae73cab8, 0xb39680, 0x7f4dae73ca30, 0xc8200b3000, 0x3, 0x3b9aca00, 0x45d964b800, 0x7ffc1f506de1, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/controller/request.go:128 +0xf5
github.com/giantswarm/inago/controller.(*controller).ExtendWithExistingSliceIDs(0xc820012e60, 0x7ffc1f506de1, 0xa, 0xb39680, 0x0, 0x0, 0xb39680, 0x0, 0x0, 0x0, ...)
    <autogenerated>:14 +0x16c
github.com/giantswarm/inago/cli.statusRun(0xb18b00, 0xc820011460, 0x1, 0x1)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/cli/status.go:38 +0x3a0
github.com/spf13/cobra.(*Command).execute(0xb18b00, 0xc820011410, 0x1, 0x1, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/spf13/cobra/command.go:569 +0x85a
github.com/spf13/cobra.(*Command).ExecuteC(0xb18700, 0xb18b00, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/spf13/cobra/command.go:656 +0x55c
github.com/spf13/cobra.(*Command).Execute(0xb18700, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/spf13/cobra/command.go:615 +0x2d
main.main()
    /usr/code/main.go:11 +0x27
```

I receive a panic when attempting to get the status of the submitted group

i also get a panic on update:

```
core@core-01 ~ $ ./inagoctl update test-group
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0x84a600, 0xc820010080)
    /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/giantswarm/inago/common.SliceID(0xc8200b34c0, 0x17, 0x0, 0x0, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/common/common.go:33 +0x137
github.com/giantswarm/inago/fleet.fleet.createOurStatusList(0xc82000ba10, 0x8d67e0, 0x4, 0x0, 0x0, 0x0, 0x8e1770, 0xb, 0x0, 0x0, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/fleet/fleet.go:314 +0x119
github.com/giantswarm/inago/fleet.fleet.GetStatusWithMatcher(0xc82000ba10, 0x8d67e0, 0x4, 0x0, 0x0, 0x0, 0x8e1770, 0xb, 0x0, 0x0, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/fleet/fleet.go:292 +0x6e9
github.com/giantswarm/inago/fleet.(*fleet).GetStatusWithMatcher(0xc82007a5a0, 0xc820012eb0, 0x0, 0x0, 0x0, 0x0, 0x0)
    <autogenerated>:7 +0xe3
github.com/giantswarm/inago/controller.controller.getExistingSliceIDs(0x7f402fc549a8, 0xc82007a5a0, 0x7f402fc54ab8, 0xb39680, 0x7f402fc54a30, 0xc8200b30e0, 0x3, 0x3b9aca00, 0x45d964b800, 0x7ffc78a71de1, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/controller/request.go:102 +0xee
github.com/giantswarm/inago/controller.controller.ExtendWithExistingSliceIDs(0x7f402fc549a8, 0xc82007a5a0, 0x7f402fc54ab8, 0xb39680, 0x7f402fc54a30, 0xc8200b30e0, 0x3, 0x3b9aca00, 0x45d964b800, 0x7ffc78a71de1, ...)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/controller/request.go:128 +0xf5
github.com/giantswarm/inago/controller.(*controller).ExtendWithExistingSliceIDs(0xc820012e60, 0x7ffc78a71de1, 0xa, 0xb39680, 0x0, 0x0, 0xc8200b3180, 0x1, 0x1, 0x0, ...)
    <autogenerated>:14 +0x16c
github.com/giantswarm/inago/cli.updateRun(0xb19100, 0xc820011460, 0x1, 0x1)
    /usr/code/.gobuild/src/github.com/giantswarm/inago/cli/update.go:55 +0x572
github.com/spf13/cobra.(*Command).execute(0xb19100, 0xc820011410, 0x1, 0x1, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/spf13/cobra/command.go:569 +0x85a
github.com/spf13/cobra.(*Command).ExecuteC(0xb18700, 0xb19100, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/spf13/cobra/command.go:656 +0x55c
github.com/spf13/cobra.(*Command).Execute(0xb18700, 0x0, 0x0)
    /usr/code/.gobuild/src/github.com/spf13/cobra/command.go:615 +0x2d
main.main()
    /usr/code/main.go:11 +0x27
```

i feel like i'm doing something wrong here

Commit:
fixes

xh3b4sd commented:
Ok I fixed some things and added error handling. Updating groups that don't have slice IDs is not possible. @JosephSalisbury please try again with a scaled group.

Commit:
Merge branch 'master' into update-command

JosephSalisbury commented:
more of me blundering around pretending to test:

this is my test set up:

```
core@core-01 ~ $ ls -R
.:
inago.service  inagoctl  test-group

./test-group:
test-group-unit@.service
core@core-01 ~ $ cat ./test-group/test-group-unit\@.service
[Unit]
Description=Inago Test Unit

[Service]
ExecStart=/bin/bash -c "while true; do echo Hello; sleep 10; done"
```

(i'm only using one vagrant machine, fyi)

```
core@core-01 ~ $ fleetctl list-units
UNIT    MACHINE ACTIVE  SUB
core@core-01 ~ $ ./inagoctl submit test-group
Succeeded to submit 1 slices for group 'test-group': [438].
core@core-01 ~ $ ./inagoctl start test-group
Succeeded to start 3 slices for group 'test-group': [2fb 3f1 438].
core@core-01 ~ $ fleetctl list-units
UNIT                MACHINE             ACTIVE  SUB
test-group-unit@2fb.service b944ede1.../172.17.8.101    active  running
test-group-unit@3f1.service b944ede1.../172.17.8.101    active  running
test-group-unit@438.service b944ede1.../172.17.8.101    active  running
```

why do i have three slices now? 

```
core@core-01 ~ $ ./inagoctl status test-group
Group           Units  FDState   FCState   SAState  IP            Machine

test-group@2fb  *      launched  launched  active   172.17.8.101  b944ede1e63146e0bf1dc7ccd01d4adb
test-group@3f1  *      launched  launched  active   172.17.8.101  b944ede1e63146e0bf1dc7ccd01d4adb
test-group@438  *      launched  launched  active   172.17.8.101  b944ede1e63146e0bf1dc7ccd01d4adb
```

shouldn't this have folded up into just 'test-group'? they're all on the same machine / ip / have the same state

```
core@core-01 ~ $ ./inagoctl submit test-group 5
Succeeded to submit 5 slices for group 'test-group': [b15 c25 99a c3e e9d].
core@core-01 ~ $ ./inagoctl start test-group
Succeeded to start 5 slices for group 'test-group': [99a b15 c25 c3e e9d].

I THEN MODIFIED test-group/test-group-unit.service

core@core-01 ~ $ ./inagoctl update test-group
Succeeded to update 4 slices for group 'test-group': [0b9 423 b9a e90].
core@core-01 ~ $ fleetctl list-units
UNIT                MACHINE             ACTIVE  SUB
test-group-unit@0b9.service b944ede1.../172.17.8.101    active  running
test-group-unit@423.service b944ede1.../172.17.8.101    active  running
test-group-unit@b9a.service b944ede1.../172.17.8.101    active  running
test-group-unit@e90.service b944ede1.../172.17.8.101    active  running
```

i appear to have lost a slice :( the remaining slices were updated correctly
i can't seem to replicate this, mind

without changing the unit file:

```
core@core-01 ~ $ ./inagoctl update test-group
Failed to update 4 slices for group 'test-group': [0b9 423 b9a e90]. (update not allowed: units already up to date)
core@core-01 ~ $ echo $?
1
```

i'm not certain i'd agree this is a failure? not certain - i'd argue we could just print 'no update required'

otherwise, i've been messing around with it, i can't see any major concerns

Commit:
cleanup

xh3b4sd commented:
Ok, thanks. So looks like there is some issue with fleet sometimes. I had the case were I had orphaned units and I saw similar weird behaviours like that. After cleaning up my cluster it worked as expected again. I fear we cannot do that much about messed up fleet nodes. 

At next I will take care about testing and stuff. To me it looks like can go with the state of this.

zeisss commented:
I have the feeling I ended up in an endless loop somewhere. At least `inagoctl update` is not doing anything to my fleet units anymore.

```
core@core-01 ~ $ time ./inagoctl update sliced --min-alive=2 --max-growth=0 --verbose
Succeeded to update 3 slices for group 'sliced': [3b0 cf8 f18].

real    0m53.068s
user    0m0.055s
[Unit]
sys 0m0.030s
core@core-01 ~ $ ./inagoctl status sliced
Group       Units  FDState   FCState   SAState  IP            Machine

sliced@3b0  *      launched  launched  active   172.17.8.103  99f23dbedf2f4bdd9ed9f2600b5944c3
sliced@cf8  *      launched  launched  active   172.17.8.102  52f3c1e0193243d8aedbf54052f83c84
sliced@f18  *      launched  launched  active   172.17.8.103  99f23dbedf2f4bdd9ed9f2600b5944c3

core@core-01 ~ $ time ./inagoctl update sliced --min-alive=2 --max-growth=0 --ready-secs=30 --verbose
Failed to update 3 slices for group 'sliced': [3b0 cf8 f18]. (update not allowed: units already up to date)

real    0m2.022s
user    0m0.001s
sys 0m0.003s
core@core-01 ~ $ vi sliced/sliced-1\@.service
core@core-01 ~ $ time ./inagoctl update sliced --min-alive=2 --max-growth=0 --ready-secs=30 --verbose
Failed to update 3 slices for group 'sliced': [7db cf8 f18]. (update failed: slice not running: [])

real    0m9.028s
user    0m0.025s
sys 0m0.012s
core@core-01 ~ $ ./inagoctl status sliced -v
Group       Units                 FDState   FCState   SAState  Hash                                      IP            Machine

sliced@7db  sliced-1@7db.service  launched  launched  active   81a5dad44645be5b8ef873f87a5ce0cbe6c11625  172.17.8.101  4ea5734c9d6347829d9236502b79c6ca
sliced@cf8  sliced-1@cf8.service  launched  launched  active   b6e536a54d2d304adda2f7554ca4fe2c07b798f8  172.17.8.102  52f3c1e0193243d8aedbf54052f83c84
sliced@f18  sliced-1@f18.service  launched  launched  active   b6e536a54d2d304adda2f7554ca4fe2c07b798f8  172.17.8.103  99f23dbedf2f4bdd9ed9f2600b5944c3
sliced@7db  sliced-2@7db.service  launched  launched  active   9272f1c566e18239338f2d12268316d47603c0d4  172.17.8.101  4ea5734c9d6347829d9236502b79c6ca
sliced@cf8  sliced-2@cf8.service  launched  launched  active   9272f1c566e18239338f2d12268316d47603c0d4  172.17.8.102  52f3c1e0193243d8aedbf54052f83c84
sliced@f18  sliced-2@f18.service  launched  launched  active   9272f1c566e18239338f2d12268316d47603c0d4  172.17.8.103  99f23dbedf2f4bdd9ed9f2600b5944c3

core@core-01 ~ $ time ./inagoctl update sliced --min-alive=2 --max-growth=0 --ready-secs=30 --verbose
^C

real    1m49.349s
user    0m0.120s
sys 0m0.072s
core@core-01 ~ $ time ./inagoctl update sliced --min-alive=2 --max-growth=0 --ready-secs=30 --verbose
^C

real    0m48.463s
user    0m0.052s
sys 0m0.032s
```

Commit:
get tests green again

Commit:
Merge branch 'master' into update-command

Commit:
tests

zeisss commented on cli/start.go:
```diff
@@ -17,7 +19,22 @@ var (
+	switch len(args) {
+	case 1:
+		group = args[0]
+	case 0:
```
I feel this is unnecessary

xh3b4sd replied:
Right. Removed.

zeisss commented on cli/start.go:
```diff
@@ -17,7 +19,22 @@ var (
+		os.Exit(1)
+	}
+
+	newRequestConfig := controller.DefaultRequestConfig()
```
3 operations against the controller to do the `start` - feels a bit overly complex. I would suggest to rethink this any maybe we can come up with sthg better (outside of this PR ofc)

xh3b4sd replied:
What is a operation in your sense? What is complex? 

We want to do the start. This is done using the controller. The controller needs a request. 

To create the request we create the request instance using its configuration. 

My goal was to have a well defined API.

zeisss replied:
I was thinking about a client/server future, and everything we define on the Controller would have an equivalent HTTP route. So if we would migrate this code to HTTP calls, we would do 3 network operations, which feels a bit too much. But we probably just have different thoughts here ;)

zeisss commented on cli/update.go:
```diff
@@ -0,0 +1,85 @@
+
+	taskObject, err := newController.Update(req, opts)
+	handleUpdateCmdError(err)
+	taskObject, err = newController.WaitForTask(taskObject.ID, nil)
```
Why do we do the `WaitForTask` here? shouldn't the `maybeBlockWithFeedback` do this already?

xh3b4sd replied:
The update creates new slices. Thus new slice IDs. I wanted to give the feedback about the new slice IDs at the end. So we need to fetch the new slice IDs once the task has finished. I don't wanted to mix this specific detail with the general implementation of `maybeBlockWithFeedback`.

zeisss replied:
Makes sense, can we add this as a comment?

xh3b4sd replied:
Done

zeisss commented on controller/controller.go:
```diff
@@ -305,6 +327,36 @@ func (c controller) Destroy(req Request) (*task.Task, error) {
+			return maskAny(err)
+		}
+
+		// TODO retry operations
```
`UpdateWithStrategy` should probably handle all the retries internally. No need for an extra TODO here IMO

xh3b4sd replied:
Not sure. This was also only copy paste. We have this in each method. Lets clean that up later.

zeisss commented on controller/request.go:
```diff
@@ -0,0 +1,200 @@
+	return r, nil
+}
+
+func (c controller) readUnitFiles(dir string) (map[string]string, error) {
```
Why does the controller now knows about the filesystem?

xh3b4sd replied:
My thoughts went into the direction of what the request need to look like. The controller requires a request. The request needs to hold certain information. Thinking about a server client architecture this makes not that much sense, but we are not there. Thus I opted for consistency for the request creation, even this is not future prove.

zeisss commented:
Lets merge it and iterate :)

Commit:
cleanup

Commit:
comment

xh3b4sd commented:
I created an issue for the `--max-growth=0` problem. See https://github.com/giantswarm/inago/issues/80
